node.go

```diff
@@ -106,15 +106,23 @@ func (n *ProtoNode) SetCidBuilder(builder cid.Builder) error {
 		n.builder = v0CidPrefix
 		return nil
 	}
-	if p, ok := builder.(*cid.Prefix); ok {
+	if p, ok := builder.(cid.Prefix); ok {
 		mhLen := p.MhLength
 		if mhLen <= 0 {
 			mhLen = -1
 		}
+		// ensure the hasher is a usable one
 		_, err := mhcore.GetVariableHasher(p.MhType, mhLen)
 		if err != nil {
 			return err
 		}
+	} else {
+		// have to test it's a usable hasher directly
+		// this is only a basic check, there are still ways it may break
+		_, err := builder.Sum([]byte{0})
```

Comment:
Please make this a global and slightly bigger (`var zeros = make([]byte, 256)`).
Making it a global avoid a trivially dodgeable allocation here.
And slightly bigger is because some algorithms wont do anything for inputs that small.